sdk/database/newdbplugin/grpc_server.go

```diff
@@ -0,0 +1,190 @@
+package newdbplugin
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/hashicorp/vault/sdk/database/newdbplugin/proto"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var _ proto.DatabaseServer = gRPCServer{}
+
+type gRPCServer struct {
+	impl Database
+}
+
+// Initialize the database plugin
+func (g gRPCServer) Initialize(ctx context.Context, request *proto.InitializeRequest) (*proto.InitializeResponse, error) {
+	rawConfig := structToMap(request.ConfigData)
+
+	dbReq := InitializeRequest{
+		Config:           rawConfig,
+		VerifyConnection: request.VerifyConnection,
+	}
+
+	dbResp, err := g.impl.Initialize(ctx, dbReq)
+	if err != nil {
+		return &proto.InitializeResponse{}, status.Errorf(codes.Internal, "failed to initialize: %s", err)
+	}
+
+	newConfig, err := mapToStruct(dbResp.Config)
+	if err != nil {
+		return &proto.InitializeResponse{}, status.Errorf(codes.Internal, "failed to marshal new config to JSON: %s", err)
+	}
+
+	resp := &proto.InitializeResponse{
+		ConfigData: newConfig,
+	}
+
+	return resp, nil
+}
+
+func (g gRPCServer) NewUser(ctx context.Context, req *proto.NewUserRequest) (*proto.NewUserResponse, error) {
+	if req.GetUsernameConfig() == nil {
+		return &proto.NewUserResponse{}, status.Errorf(codes.InvalidArgument, "missing username config")
+	}
+
+	var expiration time.Time
+
+	if req.GetExpiration() != nil {
+		exp, err := ptypes.Timestamp(req.GetExpiration())
+		if err != nil {
+			return &proto.NewUserResponse{}, status.Errorf(codes.InvalidArgument, "unable to parse expiration date: %s", err)
+		}
+		expiration = exp
+	}
+
+	dbReq := NewUserRequest{
+		UsernameConfig: UsernameMetadata{
+			DisplayName: req.GetUsernameConfig().GetDisplayName(),
+			RoleName:    req.GetUsernameConfig().GetRoleName(),
+		},
+		Password:           req.GetPassword(),
+		Expiration:         expiration,
+		Statements:         getStatementsFromProto(req.GetStatements()),
+		RollbackStatements: getStatementsFromProto(req.GetRollbackStatements()),
+	}
+
+	dbResp, err := g.impl.NewUser(ctx, dbReq)
+	if err != nil {
+		return &proto.NewUserResponse{}, status.Errorf(codes.Internal, "unable to create new user: %s", err)
+	}
+
+	resp := &proto.NewUserResponse{
+		Username: dbResp.Username,
+	}
+	return resp, nil
+}
+
+func (g gRPCServer) UpdateUser(ctx context.Context, req *proto.UpdateUserRequest) (*proto.UpdateUserResponse, error) {
+	if req.GetUsername() == "" {
+		return &proto.UpdateUserResponse{}, status.Errorf(codes.InvalidArgument, "no username provided")
+	}
+
+	dbReq, err := getUpdateUserRequest(req)
+	if err != nil {
+		return &proto.UpdateUserResponse{}, status.Errorf(codes.InvalidArgument, err.Error())
```

Comment:
Instead of returning an error, we could have `UpdateUser` return a `bool` (or even better, within `UpdateUserResponse` itself) that indicates whether this operation resulted in a no-op or not.

Reply:
Maybe this is not quite the proper approach, actually. I see why this check is implemented on the gRPC layer, but OTOH it feels that Vault usually allows for no-op updates to happen so we shouldn't be guarding by default. This could be handled in some other way, but I don't think it should be reflected as an error.

Reply:
I think this is more valuable as an error here because this should only happen if Vault is calling the plugin without any information to update. This only happens if the DB engine itself isn't providing any updated field information. I view this error as a programmer-time error, not a runtime error.

Reply:
I misunderstood this, thinking that it was a comparison check against what was already in place, but it's more of an empty check to make sure that password or expiration is provided.